pkg/karmadactl/promote.go

```diff
@@ -0,0 +1,479 @@
+	"github.com/spf13/pflag"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/cli-runtime/pkg/printers"
+	"k8s.io/cli-runtime/pkg/resource"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+
+	clusterv1alpha1 "github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1"
+	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
+	workv1alpha2 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha2"
+	karmadaclientset "github.com/karmada-io/karmada/pkg/generated/clientset/versioned"
+	"github.com/karmada-io/karmada/pkg/karmadactl/options"
+	"github.com/karmada-io/karmada/pkg/resourceinterpreter/defaultinterpreter/prune"
+	"github.com/karmada-io/karmada/pkg/util/gclient"
+	"github.com/karmada-io/karmada/pkg/util/restmapper"
+)
+
+var (
+	promoteShort = `Promote resources from legacy clusters to karmada control plane.`
+	promoteLong  = `Promote resources from legacy clusters to karmada control plane.`
+)
+
+// NewCmdPromote defines the `promote` command that promote resources from legacy clusters
+func NewCmdPromote(cmdOut io.Writer, karmadaConfig KarmadaConfig, parentCommand string) *cobra.Command {
+	opts := CommandPromoteOption{}
+	opts.JSONYamlPrintFlags = genericclioptions.NewJSONYamlPrintFlags()
+
+	cmd := &cobra.Command{
+		Use:          "promote <RESOURCE_TYPE> <RESOURCE_NAME> -n <NAME_SPACE> -c <CLUSTER_NAME>",
+		Short:        promoteShort,
+		Long:         promoteLong,
+		Example:      getPromoteExample(parentCommand),
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := opts.Complete(args); err != nil {
+				return err
+			}
+			if err := opts.Validate(); err != nil {
+				return err
+			}
+			if err := RunPromote(cmdOut, karmadaConfig, opts, args); err != nil {
+				return err
+			}
+			return nil
+		},
+	}
+
+	flag := cmd.Flags()
+	opts.AddFlags(flag)
+
+	return cmd
+}
+
+func getPromoteExample(parentCommand string) string {
+	promoteExample := `
+# Promoting deployment(default/nginx) from cluster1 to karmada.
+` + parentCommand + ` promote deployment nginx -n default -c cluster1
+
+# Promoting deployment((default/nginx) with gvk from cluster1 to karmada.
+` + parentCommand + ` promote deployment.v1.apps nginx -n default -c cluster1
+
+# Only printing the converted resource, not to create them in karmada.
+` + parentCommand + ` promote deployment nginx -n default -c cluster1 -o yaml|json
+
+# Promoting secret(default/default-token) from cluster1 to karmada.
+` + parentCommand + ` promote secret default-token -n default -c cluster1
+
+# If it's cluster of pull mode, you can use --cluster-kubeconfig and --cluster-context
```

Comment:
```suggestion
# For clusters with 'Pull' mode, use '--cluster-kubeconfig' and `--cluster-context` to specify the configuration.
```

Reply:
I think we can add one more example for pull mode clusters, as the `--cluster-context` could be omitted.

Reply:
how about to add example of --dry-run

Reply:
I don't know if it's too complicated.

Do you want `--dry-run` to dump the artifests?

Reply:
`--dry-run` is a general concept in `kubectl` and `karmadactl` that seemly need not to be a example.